### ISSUE TITLE
EngineTest.ScenarioStreamRedirect fails with a LookupException

### ISSUE THREAD

**Issue by slozier** [IronLanguages/ironpython3#906]
It looks like this is due to the the standard library not being available and so none of the codecs are available via `_codecs.lookup`. The failure occurs even though the particular encoding causing the failure (cp437) is available to use via `bytes.decode`/`str.encode`. This is because `io.TextIOWrapper` does not make use of the `StringOps.TryGetEncoding` mechanism but instead relies on `PythonOps.LookupEncoding`.

Related to https://github.com/IronLanguages/ironpython2/issues/659 (probably the same issue)

@BCSharp not sure if this something you encountered before?

**Comment by BCSharp** [comment]
Yes, I was aware of that. This is one of the main reasons why several tests at the top of `test_codecs_stdlib` are not passing yet. I am not quite sure what is the best course of action though. Here are some of my musings on this topic.

1. Python documentation of `codecs` reads:
   > While the builtin `open()` and the associated `io` module are the recommended approach for working with encoded text files, this module provides additional utility functions and classes that allow the use of a wider range of codecs when working with binary files.

   This tells me that there is not much need to use text encodings in `codecs` directly, as long as `open()` and `io` are optimized to use .NET encodings, which should be the case anyway for performance reasons. `codecs` seems to me a legacy from Python 2.

2. But from the compatibility point of view, the `codecs` module should be available and complete. The quickest way to ensure this is to provide a proxy class to any .NET `Encoding` that can impersonate a Python codec by providing a `codecs`-like interface (`codecs.CodecInfo`, `codecs.IncrementalEncoder`, `codecs.IncrementalDecoder`, `codecs.StreamReader`, `codecs.StreamWriter`). And if not "the quickest" then at least the most comprehensive. There are some challenges with this approach (e.g. preserving the   state of incremental encoding/decoding is done differently in .NET than Python), but once done, all .NET codecs become available through `codecs`. Remaining codecs written in pure Python top it up and there is only a small number of codecs left out that would have to be manually addressed.

3. However, a few of .NET codecs behave slightly differently than the corresponding Python codecs, which makes me hesitant to fully embrace the above approach. So far I have noticed the discrepancies in UTF-7 and a few charmap codecs (I am ignoring discrepancies due to bugs in .NET for now). The UTF-7 case is probably acceptable, for two reasons: (1) both implementations conform to the standard and are correct, meaning that output produced by the .NET encoder can be decoded by CPython and vice versa; (2) CPython's UTF-7 codecs is written in C, with `utf_7.py` providing just a Python proxy, so unless IronPython reimplements the complete codec in C#, the .NET behaviour is as good as it gets.

   The charmap encodings story is a little bit different. In come codepages, some byte values do not have defined mapping to characters. Encountering such bytes in the input, the CPython codec will raise a decoding error. The .NET encoding will (usually? always?) map it successfully to a character using Latin-1. This also works during encoding. Charmap encodings (except `charmap` itself) are pure Python so using `codecs` guarantees the exact CPython behaviour, in case somebody needs it. On the other hand, it could be surprising to have a different behaviour between, say:

       b'\x81'.decode('windows-1252')

   and

       codecs.decode(b'\x81', 'windows-1252')

    as it is now. Also, charmap encodings from `codecs` are significantly slower than their .NET counterparts (except for Latin-1, of course).

4. An alternative approach would be to fully implement `codecs` the Python way and redirect `str.encode` and `bytes.decode` to use them to avoid discrepancies. This would require some effort as many non-trivial codecs depend on C/C# modules for the actual workload. The effort in this direction seems to have started in the past (there is a skeleton `_codecs_cn.cs` and `_multibytecodecs.cs`, for example). Besides the effort, the drawback is a performance hit.

5. But unless Python StdLib gets patched, this may not be enough: many codec proxies written in Python assume that Python strings are in UTF-32. Since in IronPython strings are UTF-16, incremental encoding will stay broken across a wide range of codecs (as it is now). Ironically, some of the remaining functionality of `codecs` that is **not** obsolete in Python 3 is support of incremental encoding/decoding.

6. "unless Python StdLib gets patched" => or unless #252 gets implemented.

Currently I am leaning towards the approach described in point 2 (with all its caveats), but since there are still other loose ends in the encoding support, I pursue them first and maybe more clarity will emerge later. A discussion could help too.
